# PART A

A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [41]:
import keras
from keras.models import Sequential
from keras.layers import Dense
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

In [4]:
# The code was removed by Watson Studio for sharing.

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [42]:
X = concrete_data.drop('Strength', axis = 1)
y = concrete_data['Strength']

mse_list = []
for i in range(50):
    print(f"Starting the {i+1}. run.")
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)
    
    model = Sequential()
    n_cols = X.shape[1]
    
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(1))
    
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, epochs = 50)
    
    predictions = model.predict(X_test)
    
    mse = mean_squared_error(y_test, predictions)
    print(f"The MSE of the model is {mse} on {i+1}. run.")
    mse_list.append(mse)

Starting the 1. run.
Epoch 1/50
23/23 [==============================] - 0s 3ms/step - loss: 23974.9316
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 5194.4854
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 2464.2522
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 2310.4358
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 2169.9087
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 2036.2694
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1898.8920
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1765.0751
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 1635.5016
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1510.9039
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1383.9482
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 1

In [35]:
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"The mean of the mean squared errors is {mean_mse}.")
print(f"The standard deviation of the mean squared errors is {std_mse}.")

The mean of the mean squared errors is 198.07862949889724.
The standard deviation of the mean squared errors is 153.50957093056115.


# PART B

B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [43]:
X = concrete_data.drop('Strength', axis = 1)
y = concrete_data['Strength']

mse_list = []
for i in range(50):
    print(f"Starting the {i+1}. run.")
    
    # We perform the train-test split here.
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

    # We perform the normalization here.
    scaler = StandardScaler()
    # We will fit the scaler on the training data and use the learned scaling parameters on the test set to transform it as well.
    scaler.fit(X_train)

    # We normalize the training and test set using learned parameters from the train set.
    X_train = scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    
    n_cols = X.shape[1]
    
    model = Sequential()
    
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(1))
    
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, epochs = 50)
    
    predictions = model.predict(X_test)
    
    mse = mean_squared_error(y_test, predictions)
    print(f"The MSE of the model is {mse} on {i+1}. run.")
    mse_list.append(mse)

Starting the 1. run.
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1610.3605
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1595.8473
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1581.5879
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1567.4290
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1553.3032
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1538.4786
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1523.5232
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1507.8689
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1491.6254
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1474.6942
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1457.0236
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 14

Reporting the error metrics for the case with normalized data (Part B))

In [44]:
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"The mean of the mean squared errors is {mean_mse}.")
print(f"The standard deviation of the mean squared errors is {std_mse}.")

The mean of the mean squared errors is 353.84610558693674.
The standard deviation of the mean squared errors is 76.9554358143334.


The mean of the mean squared errors was *198.08* and the standard deviation was *153.51* in part A, where the data was not normalized.

In part B, we see that the mean is up to *353.85* and the standard deviation is down to *76.95*. While the mean error greatly increased, there is a decrease in the standard deviation.

The decrease in the standard deviation is expected since we performed normalization on the data and we hopefully expect to eliminate the impact of outliers in some features of the data. On the other hand, we the mean error's skyrocketed, and the reason for that may be an underfitting problem. One solution can be to increase the number of epochs.

# PART C

C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [45]:
X = concrete_data.drop('Strength', axis = 1)
y = concrete_data['Strength']

mse_list = []
for i in range(50):
    print(f"Starting the {i+1}. run.")
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

    # We perform the normalization here.
    scaler = StandardScaler()
    # We will fit the scaler on the training data and use the learned scaling parameters on the test set to transform it as well.
    scaler.fit(X_train)

    # We normalize the training and test set using learned parameters from the train set.
    X_train = scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    
    n_cols = X.shape[1]
    
    model = Sequential()
    
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(1))
    
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, epochs = 100)
    
    predictions = model.predict(X_test)
    
    mse = mean_squared_error(y_test, predictions)
    print(f"The MSE of the model is {mse} on {i+1}. run.")
    mse_list.append(mse)

Starting the 1. run.
Epoch 1/100
23/23 [==============================] - 0s 2ms/step - loss: 1614.6183
Epoch 2/100
23/23 [==============================] - 0s 2ms/step - loss: 1602.6101
Epoch 3/100
23/23 [==============================] - 0s 2ms/step - loss: 1590.4402
Epoch 4/100
23/23 [==============================] - 0s 2ms/step - loss: 1578.0641
Epoch 5/100
23/23 [==============================] - 0s 2ms/step - loss: 1565.0536
Epoch 6/100
23/23 [==============================] - 0s 2ms/step - loss: 1551.3811
Epoch 7/100
23/23 [==============================] - 0s 2ms/step - loss: 1536.7875
Epoch 8/100
23/23 [==============================] - 0s 2ms/step - loss: 1521.0444
Epoch 9/100
23/23 [==============================] - 0s 2ms/step - loss: 1504.2755
Epoch 10/100
23/23 [==============================] - 0s 2ms/step - loss: 1486.0653
Epoch 11/100
23/23 [==============================] - 0s 2ms/step - loss: 1466.4131
Epoch 12/100
23/23 [==============================] - 0s 2ms/ste

In [46]:
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"The mean of the mean squared errors is {mean_mse}.")
print(f"The standard deviation of the mean squared errors is {std_mse}.")

The mean of the mean squared errors is 162.44407896344956.
The standard deviation of the mean squared errors is 12.632896661280475.


As we expected, both mean and standard deviation of the mean squared errors lowered significantly when we increased the epochs. The model learned better, so the performance improved.

# PART D

D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [47]:
X = concrete_data.drop('Strength', axis = 1)
y = concrete_data['Strength']

mse_list = []
for i in range(50):
    print(f"Starting the {i+1}. run.")
    
    # We perform the train-test split here.
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

    # We perform the normalization here.
    scaler = StandardScaler()
    # We will fit the scaler on the training data and use the learned scaling parameters on the test set to transform it as well.
    scaler.fit(X_train)

    # We normalize the training and test set using learned parameters from the train set.
    X_train = scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    
    n_cols = X.shape[1]
    
    model = Sequential()
    
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(1))
    
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, epochs = 50)
    
    predictions = model.predict(X_test)
    
    mse = mean_squared_error(y_test, predictions)
    print(f"The MSE of the model is {mse} on {i+1}. run.")
    mse_list.append(mse)

Starting the 1. run.
Epoch 1/50
23/23 [==============================] - 0s 4ms/step - loss: 1578.8641
Epoch 2/50
23/23 [==============================] - 0s 3ms/step - loss: 1558.3408
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1544.0127
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1533.7058
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1521.7751
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1503.2513
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1474.1947
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1429.5826
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1368.1023
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1285.0283
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1174.3293
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 10

In [48]:
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"The mean of the mean squared errors is {mean_mse}.")
print(f"The standard deviation of the mean squared errors is {std_mse}.")

The mean of the mean squared errors is 128.99052196809876.
The standard deviation of the mean squared errors is 12.762537360432885.


In part B, the mean was *353.85* and the standard deviation of the mean squared errors was *76.95*.

When we changed the model's architecture to 3 hidden layers in part D, the mean dropped to *128.99* and the standard deviation lowered to *12.76*.

We can conclude that increasing the complexity of the model also helped the performance of the model.